### PR TITLE
New methods requested for ssh port forwarding

### PR DESCRIPTION
**Pull request by robertodeandrea**
These are the minor changes required for integrate actual port forwarding mechanism in my code.

### CODE REVIEW

File: sshd-core/src/main/java/org/apache/sshd/client/session/AbstractClientSession.java [hunk 1]
```diff
@@ -330,6 +330,18 @@ public void stopDynamicPortForwarding(SshdSocketAddress local) throws IOExceptio
...
+        ForwardingFilter filter = getForwardingFilter();
+        return filter.hasLocalPortForwardingStarted(local);
+    }
+
+    @Override
+    public boolean hasRemotePortForwardingStarted(SshdSocketAddress remote) {
+        ForwardingFilter filter = getForwardingFilter();
+        return filter.hasRemotePortForwardingStarted(remote);
+    }
+
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/client/session/AbstractClientSession.java]: Please be _null_-tolerant:
```java
ForwardingFilter filter = getForwardingFilter();
return (filter != null) && filter.hasXXXX();
```

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 2]
```diff
@@ -1017,4 +1017,34 @@ public void exceptionCaught(IoSession session, Throwable cause) throws Exception
...
+        ValidateUtils.checkTrue(remote.getPort() >= 0, "Invalid remote port: %s", remote);
+    
+        synchronized (remoteToLocal) {
+            SshdSocketAddress local = remoteToLocal.get(Integer.valueOf(remote.getPort()));
+            if (local != null) {
+                started = true;  
+            }
+        }      
+        return started;
+    }
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: No need to use a separate variable for the return - do it a.s.a.p:
```java
synchronized (xxx) {
    return xxx.containsKey(...port...);
}
```

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 3]
```diff
@@ -1017,4 +1017,34 @@ public void exceptionCaught(IoSession session, Throwable cause) throws Exception
             session.close(true);
         }
     }
+
+    @Override
+    public boolean hasLocalPortForwardingStarted(SshdSocketAddress local) {
+        boolean started = false;
+        Objects.requireNonNull(local, "Local address is null");
+        ValidateUtils.checkTrue(local.getPort() >= 0, "Invalid local port: %s", local);
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: Shouldn't this be `local.getPort() > 0` - what is the meaning of port=_0_ ?

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 4]
```diff
@@ -1017,4 +1017,34 @@ public void exceptionCaught(IoSession session, Throwable cause) throws Exception
...
+            }
+        }      
+        return started;
+    }
+
+    @Override
+    public boolean hasRemotePortForwardingStarted(SshdSocketAddress remote) {
+        boolean started = false;
+        Objects.requireNonNull(remote, "Remote address is null");
+        ValidateUtils.checkTrue(remote.getPort() >= 0, "Invalid remote port: %s", remote);
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: Shouldn't this be `local.getPort() > 0` - what is the meaning of port=_0_ ?

File: sshd-core/src/main/java/org/apache/sshd/client/session/AbstractClientSession.java (hunk 1 above)
Comment (robertodeandrea) [reply sshd-core/src/main/java/org/apache/sshd/client/session/AbstractClientSession.java]: ok, I agree

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (robertodeandrea) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: ok; I agree

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 3 above)
Comment (robertodeandrea) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: ok, I agree

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 4 above)
Comment (robertodeandrea) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: ok,I agree

**Comment by gnodet** [comment]
Merged

**Pull request closed without merging**